libraries/AP_Airspeed/AP_Airspeed.cpp

```diff
@@ -399,15 +427,132 @@ void AP_Airspeed::read(void)
     }
 #endif
 
-    // setup primary
-    if (healthy(primary_sensor.get())) {
-        primary = primary_sensor.get();
+    if (blend_state.enable == BLEND_NONE ||
+        !AP::ahrs().have_inertial_nav()) {
+        process_no_blend();
         return;
     }
-    for (uint8_t i=0; i<AIRSPEED_MAX_SENSORS; i++) {
-        if (healthy(i)) {
-            primary = i;
-            break;
+    
+    // blend the airspeed from several sensors and EKF
+    Vector3f velocity, wind;
+    if (AP::ahrs().get_velocity_NED(velocity) == false) {
+        process_no_blend();
+        return;
+    }
+    wind = AP::ahrs().wind_estimate();
+    velocity -= wind;
+    float airspeed_ekf = velocity.length();
+
+    float airspeed_avg = 0.0;
+    float ekf_sensor_airspeed_diff[AIRSPEED_MAX_SENSORS];
+    uint8_t healthy_sens_num = 0;
+    for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+        if (!state[i].healthy) {
+            state[i].blend_healthy = false;
+        }
+        
+        if (state[i].blend_healthy) {
+            airspeed_avg += state[i].airspeed;
+            healthy_sens_num++;
+        }
+    }
+    
+    // calculating difference
+    for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+        ekf_sensor_airspeed_diff[i] = fabsf(state[i].airspeed - airspeed_ekf);
+    }
+    
+    if (healthy_sens_num > 0)
+    {
+        airspeed_avg /= healthy_sens_num;
+        
+        const float airspeed_avg_limit = airspeed_ekf * AIRSPEED_BLEND_DIFF_LIMIT;
+
+        // look for healthy sensors that might be failing
+
+        // Idea is to find the sensor which differs much both from other sensors and from ekf value
+        // If there are two sensors then the first check will be true for both, however
+        // the second check will be true only for the sensor which really failed
+        for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+            if (state[i].blend_healthy &&
+                    ekf_sensor_airspeed_diff[i] > airspeed_avg_limit && state[i].airspeed > AIRSPEED_BLEND_MIN_SPEED) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Airspeed[%u] sensor blending unhealthy", i);
+                state[i].blend_healthy = false;
+            }
+        }
+        
+        blend_state_t blend_state_temp;
+        memcpy(&blend_state_temp, &blend_state, sizeof(blend_state));
```

Comment:
don't memcpy here, you just zero it afterwards anyway. Use:
`blend_state_t blend_state_temp {}; // the {} memsets to zero`

here's an example:
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/GCS_Mavlink.cpp#L897

Reply:
Not all members should be zeroed.
AP_Int8 enable should not be for sure.

Reply:
good call. I totally missed that. What idiot told you to include that in the struct! :)

Reply:
Before memcpy that did not hurt

Reply:
AP_Int8 enable left the structure for good